content/posts/numpy/generator-rng/index.md

```diff
@@ -0,0 +1,304 @@
+]
+
+# visualization
+fig, axs = plt.subplots(1, 3, figsize=(12, 4))
+
+for i, dist in enumerate(dists):
+    _, bins, _ = axs[i].hist(dist.sample, bins=64, density=True)
+    x = np.linspace(min(bins), max(bins), 100)
+    axs[i].plot(x, dist.pdf.pdf(x), 'r-', lw=5, alpha=0.6, label=dist.name)
+    axs[i].legend()
+```
+
+![
+Histograms for uniform, normal and gamma distribution using 1024 samples.
+They are compared with the underlying distributions and a poor match. 
+](distributions_seeded.png)
+
+Does that look good to you? It is not. If you need to have a determined
+sequence of points, there are other technics that would generate almost
+perfect histograms only using a fraction of the number of points which is used
+here. See [this article]({{< relref "../../scipy/qmc-basics" >}})
+for more information on this.
+
+But why does it matter? How bad this is will depend on your application.
+If you don't, statistical properties are lost. This is hard to get because for
+most people, random is random. So taking an arbitrary draw of sample should not
+matter. And it does not **if** you replicate your analysis using different
+**random** seeds. This is the part everyone is forgetting. Whether you are
+optimizing something or calculating some statistics, you could get (un)lucky
+if you do this just once.
+
+If your application needs random numbers, then do use random numbers and play
+with the fact that they are random. You do science or anything critical
+using these randomly generated numbers, then please stay away from seeds.
+
+Some will argue that seeding is just used by developers for testing, well it is
+not. People don't like uncertainties. They naturally try to "fix" this
+"issue" by using a seed in their systems, and not just for testing.
+The worst part here is that people tend to use the same numbers. Some
+copy/pasting might be at play there as well.
+
+Think about it for a while. If everybody is using a seed, which is the same,
+and the original intent was to interact with random numbers for statistical
+analysis or optimization. Do you see the bias that is being introduced?
+
+## Testing with random numbers
+
+Before showing any code though, ask yourself if you actually need ot use
+random numbers to write your test. You might just need a small matrix with
+hard-coded numbers. Not only you don't need to run the code to know the values,
+but you will be 100% certain that this is stable on any platform/architecture
+and future-proof. You have more advanced needs, ok then maybe using known
+sequences (quasi-random numbers) is a good alternative (again see
+[here]({{< relref "../../scipy/qmc-basics" >}})
+).
+
+If you want or need to use NumPy for that, then either use
+`np.random.Generator` or the legacy `np.random.RandomState`. Aside from
+backward compatibility concerns, `np.random.RandomState` is still present in
+NumPy for testing.
+
+```python
+rng = np.random.default_rng()
```

Comment:
`default_rng()` is not necessarily stable.  So I think for testing `RandomState` should be used?

Reply:
Agree.  Nothing in `default_rng` should be considered long-termstable, even the  `BitGenerator`

Reply:
I added clearer comments